### ISSUE TITLE
FastFlix occasionally getting stuck on a single video in a queue

### ISSUE THREAD

**Issue by kamild1996** [cdgriffith/FastFlix#278]
It has happened to me twice now and I'm not sure what's the cause.
I'm currently using FastFlix (v4.5.1) to mass convert about 5-10 videos every night. However, it seems like FastFlix likes to get stuck on one random video in the queue, continuously encoding it over and over (overwriting the output file every time) until I stop the encoding process. The first time it was the second video from 8 in the queue, the second time it was the last out of 5.

What I have also noticed is that the video that got looped shows up twice in the queue, even though I'm certain it hasn't been present twice in the queue at the moment when I started the encoding process.

**Comment by cdgriffith** [comment]
Sorry for delay, holidays! That is very weird, and news to me so thanks for bringing it up. Could you attach or email me your logs for those times? 

In FastFlix go to Help > Open Log Directory. Can either comb through them or just zip the directory. Feel free to sanitize them as well (they will include full paths to videos being converted, as well as paths to tools. So may want to remove username or video names.)

**Comment by jeremysgates** [comment]
> It has happened to me twice now and I'm not sure what's the cause. I'm currently using FastFlix (v4.5.1) to mass convert about 5-10 videos every night. However, it seems like FastFlix likes to get stuck on one random video in the queue, continuously encoding it over and over (overwriting the output file every time) until I stop the encoding process. The first time it was the second video from 8 in the queue, the second time it was the last out of 5.
> 
> What I have also noticed is that the video that got looped shows up twice in the queue, even though I'm certain it hasn't been present twice in the queue at the moment when I started the encoding process.

I wanted to report that I am having a similar issue as well. However, I don't believe the process actually gets stuck, it just looks like it does from the GUI/queue.  For me it seems to happen after adding several additional movies to the queue to encode after I have an active encode running.  The movie that is actively encoding in the "Queue" tab no longer shows up in the queue, but if you go to the "Encoding Status" tab, you can see the remaining encode time counter looks about right.  If you leave it alone and let it finish the current encode (by simply monitoring the remaining encode time in the "Encoding Status" tab), you'll find that the movie does successfully finish encoding as expected.  You just don't get any notification of it in the GUI application.  

Additionally, the next movie in the queue looks to be correct, but the one queued after it becomes duplicated and displays as the following 2 queue items in the "Encoding Queue" tab.

So to summarize:

- Currently encoding item disappears from the queue
- Next item is queue is correctly displayed
- Subsequent two items are duplicated

Here's a link to my fastflix logs folder, queue.yaml, and screenshot of my queue in the fastflix GUI.  I hope this information helps.  Just let me know if any additional info is needed and thanks for such a great application!

https://drive.google.com/file/d/1Myz6Ys_sNAJbKi3F0byp9WTolpdWz_Bf/view?usp=sharing

**Comment by cdgriffith** [comment]
Thank you @jeremysgates for the write up and examples! Will probably take me a bit to figure out what synchronization issue is causing this, but is a big bug so will need to get too it!

**Comment by jeremysgates** [comment]
> 

@cdgriffith ,  
Providing some additional details that might help isolate the issue.  The issue seems to start occurring when adding a 4th item to the encode queue (at least for me).  Once that 4th item is added to the queue, the currently encoding job updates to show "Encode Completed" status in the "Encoding Queue" tab.  However, that active encode continues to run as expected.  Here's another screenshot example:

https://drive.google.com/file/d/1F9a1J08UPQ0oH4PTzw4fx_NdWKuIlI4a/view?usp=sharing

**Comment by cdgriffith** [comment]
I keep trying to replicate this and just don't get (un)lucky enough to have the issue. I'm guessing it's just some sync issues between the background process for encoding and the main process. I am trying to add a few locks with next bugfix release to see if it helps, but not feeling totally confident in it if I can't reproduce.

**Comment by cdgriffith** [comment]
Could you try out https://github.com/cdgriffith/FastFlix/releases/tag/4.6.0 and see if the issue still happens in the same way? I cannot reproduce it myself, so guessing for the fix. Thanks!

**Comment by jeremysgates** [comment]
@cdgriffith ,  I'll give it a go and update here. Thanks!

**Comment by cdgriffith** [comment]
I moved to a much simpler internal queue system. Basically instead of both the command runner and the frontend both having some state management of the queue itself, the command runner is now "dumb" and just run what is told by the UI.

Should make this problem go away, if there are any more issues please re-open or recreate if it's different! https://github.com/cdgriffith/FastFlix/releases/tag/4.9.0
